### ISSUE TITLE
Correct index for distance matrix

### ISSUE THREAD

**Issue by TeodoraB21** [google/or-tools#1898]
Hello!
I have a problem with 9 stops and 3 vehicles.
start for each vehicle: 0 0 1
end for each vehicle: 3 2 1
 to set the distance I have:
```
distCallback = routing->RegisterTransitCallback(
			[this](int64 from_index, int64 to_index) -> int64 {
				auto from_node = manager->IndexToNode(from_index);
				auto to_node = manager->IndexToNode(to_index);
				return m_distancesMatrix[from_node.value()][to_node.value()];
			});
```

My question is, should I return the m_distancesMatrix[from_node.value()][to_node.value()] or m_distancesMatrix[from_index][to_index]?
In the manager, the `index_to_node` looks like:
```
0 -> 0 // start vehicle 0
1 -> 1 // start vehicle 2
2 -> 4
3 -> 5
4 -> 6
5 -> 7
6 -> 8
7 -> 0 // start vehicle 1 (using the first available index)
8 ->3 // end vehicle 0
9 -> 2 // end vehicle 1
10 ->1 // end vehicle 2
```
and `node_to_index` looks like:
```
0 -> 0 // you must use StartIndex(0) or StartIndex(1), should have returned -1
1 -> 1 // you must use StartIndex(2) or EndIndex(2), should have returned -1
2 -> -1 // you must use EndIndex(1)
3 -> -1 // you must use EndIndex(0)
4 -> 2
5 -> 3
6 -> 4
7 -> 5
8 -> 6
```
So if from_index=4 and to_index=3 if i return:
- m_distancesMatrix[from_index][to_index] = m_distancesMatrix[4][3]
- m_distancesMatrix[from_node.value()][to_node.value()] = m_distancesMatrix[6][5] (because manager->IndexToNode(4) = node with value 6 and manager->IndexToNode(3) = node with value 5)

**Comment by Mizux** [comment]
index, are internal indices use by the routing solver IN/OUT so when the solver call your transit callback, the solver pass index (which have no meaning for you).

In **your** data model "world" you use node index that you feed to the RoutingIndexManager.
so using this manager you can do the conversion:
* From solver index aka index -> To your index aka NodeIndex using `manager.IndexToNode()`

note: node index `0` is ambiguous for the solver since in your model it's the start node for vehicle 0 and 1 while solver need a unique index, that's why it create a index 7 (which map also to your node 0)

TLDR:
* only use NodeToIndex for location not for Start/End nodes
* you should use IndexToNode in your callback to map solver index to your model "node index" based.
* Use `StartIndex()`/`EndIndex()` (from `RoutingIndexManager` or `Start()`/`End()` from `RoutingModel`)

If I were you, I would use:
```cpp
distCallback = routing->RegisterTransitCallback(
			[this](int64 from_index, int64 to_index) -> int64 {
				int from_node = manager->IndexToNode(from_index).value();
				int to_node = manager->IndexToNode(to_index).value();
				return m_distancesMatrix[from_node][to_node];
			});
```

**Comment by TeodoraB21** [comment]
And if I want that nodes 8,4,3 to be visited by the same vehicle, I do this:
```
int32 prevStop = m_stopsWithSameVehicle[i];
VRPNodeIndex prevNode(prevStop);
int32 prevStopIndex = manager->NodeToIndex(prevNode);

int32 stop = m_stopsWithSameVehicle[i+1];
VRPNodeIndex node(stop);
int64 nodeIndex = manager->NodeToIndex(node);

solver->AddConstraint(solver->MakeEquality(routing->VehicleVar(prevStopIndex), routing->VehicleVar(nodeIndex)));
```
Is it correct?

**Comment by Mizux** [comment]
yes !

**Comment by TeodoraB21** [comment]
but 3 is the end stop for the first vehicle, so nodeIndex = -1 and i receive an error here:
`solver->AddConstraint(solver->MakeEquality(routing->VehicleVar(prevStopIndex), routing->VehicleVar(nodeIndex)));`

**Comment by Mizux** [comment]
don't understand everything in your snippet but if one of your node is a Start/Stop you can't use `NodeToIndex()` but you need to use:
https://github.com/google/or-tools/blob/2c67e00f3a8a861440f0b8016e209c51990a0403/ortools/constraint_solver/routing_index_manager.h#L63-L64
